presto-main/src/main/resources/webapp/assets/query.js

```diff
@@ -0,0 +1,1087 @@
+            tooltipValueLookups['offset'][i] = numberFormatter(dataMin + (i * bucketSize)) + "-" + numberFormatter(dataMin + ((i + 1) * bucketSize));
+        }
+
+        var stageHistogramProperties = $.extend({}, HISTOGRAM_PROPERTIES,  {barWidth: (HISTOGRAM_WIDTH / histogramData.length), tooltipValueLookups: tooltipValueLookups});
+        $(histogramId).sparkline(histogramData, stageHistogramProperties);
+    },
+    componentDidUpdate: function() {
+        var s = this.props.stage;
+        var numTasks = s.tasks.length;
+
+        // sort the x-axis
+        s.tasks.sort(function (taskA, taskB) {
+            return getTaskIdInStage(taskA.taskStatus.taskId) - getTaskIdInStage(taskB.taskStatus.taskId);
+        })
+
+        var outgoingBuffers = s.tasks.map(function(task) {
+            return task.outputBuffers.totalBufferedBytes;
+        });
+
+        var cpuTimes = s.tasks.map(function(task) {
+            return parseDuration(task.stats.totalCpuTime);
+        });
+
+        // prevent multiple calls to setState within the refresh interval from re-rendering sparklines/charts
+        if (this.state.lastRender == null || (Date.now() - this.state.lastRender) >= 1000) {
+            var renderTimestamp = Date.now();
+            var stageHistogramProperties = $.extend({}, HISTOGRAM_PROPERTIES,  {barWidth: (HISTOGRAM_WIDTH / (Math.min(numTasks, HISTOGRAM_WIDTH) + 1))});
+
+            this.renderHistogram('#outgoing-buffer-histogram-' + s.plan.id, outgoingBuffers, formatDataSize);
+            this.renderHistogram('#cpu-time-histogram-' + s.plan.id, cpuTimes, formatDuration);
+
+            if (this.state.expanded) {
+                // this needs to be a string otherwise it will also be passed to numberFormatter
+                var tooltipValueLookups = {'offset' : {}};
+                for (var i = 0; i < numTasks; i++) {
+                    tooltipValueLookups['offset'][i] = s.plan.id + "." + i;
+                }
+
+                var stageBarChartProperties = $.extend({}, BAR_CHART_PROPERTIES, {barWidth: BAR_CHART_WIDTH / numTasks, tooltipValueLookups: tooltipValueLookups});
+
+                $('#outgoing-buffer-bar-chart-' + s.plan.id).sparkline(outgoingBuffers, $.extend({}, stageBarChartProperties, {numberFormatter: formatDataSize}));
+                $('#cpu-time-bar-chart-' + s.plan.id).sparkline(cpuTimes, $.extend({}, stageBarChartProperties, {numberFormatter: formatDuration}));
+            }
+
+            this.setState({
+                lastRender: renderTimestamp
+            });
+        }
+
+        $('[data-toggle="tooltip"]').tooltip();
+    },
+    render: function() {
+        var s = this.props.stage;
+
+        var totalBufferedBytes = s.tasks.map(function(task) {
+            return task.outputBuffers.totalBufferedBytes;
+        }).reduce(function(previousValue, currentValue) {
+            return previousValue + currentValue;
+        });
+
+        return (
```

Comment:
This block of HTML is pretty long. Maybe there's a way to split it up a bit more?